### PR TITLE
feat(goal_planner): divide Planners to isolated threads

### PR DESCRIPTION
**Pull request by soblin**
## Description

Currently `thread_safe_data_` member variable is accessed from many member functions. This structure is not healthy because there are other possible not thread-safe member variables, and numerous number of call to lock() is very insufficient.

After this PR, there are 3 agents
- goal_planner_module
- LaneParkingPlannerThread
- FreespaceParkingThread

### goal_planner_module

Lock the `lane_parking_mutex_`, and
- prepare the `LaneParkingRequest` member
- copy the `LaneParkingResponse` to local scope

Lock the `freespace_parking_mutex_`, and
- prepare the `FreespaceParkingRequest` member
- copy the `FreespaceParkingResponse` to local scope

### LaneParkingThread

Lock the `lane_parking_mutex_` which is shared with goal_planner_module via reference, and
- copy `LaneParkingResponse` to local scope

After computation,
Lock the `lane_parking_mutex_`, and
- set `LaneParkingResponse`

This will be used by goal_planner_module

### FreespaceParkingThread

Lock the `freespace_parking_mutex_` which is shared with goal_planner_module via reference, and
- copy `FreespaceParkingResponse` to local scope

After computation,
Lock the `freespace_parking_mutex_`, and
- set `FreespaceParkingResponse`

This will be used by goal_planner_module

Overall, thread  safe data has been divided to above 4 classes, and the `response` value is copied to module side  `PullOverContextData` member, so repetitive access to thread_safe_data with `lock` have been purged.

## Related links

**Parent Issue:**

- https://tier4.atlassian.net/browse/RT1-7966

## How was this PR tested?

- https://evaluation.tier4.jp/evaluation/reports/f7faf106-25c6-58ee-842c-2b5525b536be?project_id=prd_jt
- https://evaluation.tier4.jp/evaluation/reports/64a5c824-baed-50db-8ff6-6b84809ff07c?project_id=prd_jt

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/goal_planner_module.hpp [hunk 1]
```diff
@@ -217,48 +292,16 @@ class GoalPlannerModule : public SceneModuleInterface
...
-      const ModuleStatus & current_status, const BehaviorModuleOutput & previous_module_output,
-      const autoware::universe_utils::LinearRing2d & vehicle_footprint,
-      const GoalCandidates & goal_candidates);
-
-  private:
-    void initializeOccupancyGridMap(
-      const PlannerData & planner_data, const GoalPlannerParameters & parameters);
-  };
-  std::optional<GoalPlannerData> gp_planner_data_{std::nullopt};
-  std::mutex gp_planner_data_mutex_;
```
Comment (kosuke55) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/goal_planner_module.hpp:261]: `gp_planner_data` is deleted but still remain in comments

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/goal_planner_module.hpp [hunk 2]
```diff
@@ -150,6 +170,61 @@ class ScopedFlag
...
+  void onTimer();
+
+private:
+  std::mutex & mutex_;
+  const std::optional<LaneParkingRequest> & request_;
+  LaneParkingResponse & response_;
+  std::atomic<bool> & is_lane_parking_cb_running_;
+  rclcpp::Logger logger_;
+
+  std::vector<std::shared_ptr<PullOverPlannerBase>> pull_over_planners_;
```
Comment (kosuke55) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/goal_planner_module.hpp:191]: nits:
(I know this is just using previous code. but we can use unique_ptr or just vector. I want refactor in the future)

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 3]
```diff
@@ -267,50 +280,72 @@ bool isStopped(
...
+        node, parameters, lane_departure_checker, /*is_forward*/ true));
+    } else if (planner_type == "ARC_BACKWARD" && parameters.enable_arc_backward_parking) {
+      pull_over_planners_.push_back(std::make_shared<GeometricPullOver>(
+        node, parameters, lane_departure_checker, /*is_forward*/ false));
+    }
+  }
+
+  if (pull_over_planners_.empty()) {
+    RCLCPP_ERROR(logger_, "Not found enabled planner");
+  }
```
Comment (kosuke55) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:317]:
```suggestion
  if (pull_over_planners_.empty()) {
    RCLCPP_WARN(
      getLogger(),
      "No enabled planner found. The vehicle will stop in the road lane without pull over. Please "
      "check the parameters if this is the intended behavior.");
  }
```

currently I changed some log messages in https://github.com/autowarefoundation/autoware.universe/pull/9562

there may be similar parts, so please check

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 4]
```diff
@@ -670,6 +659,7 @@ void GoalPlannerModule::updateData()
     goal_candidates_ = generateGoalCandidates();
   }
 
+  occupancy_grid_map_->setMap(*(planner_data_->occupancy_grid));
```
Comment (kosuke55) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:662]: memo for me.
it is ok not to lock becauses this is used in only main thread.

Comment (kosuke55) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:662]: currently it is used in isStuck(), but will remove that

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 5]
```diff
@@ -118,19 +91,30 @@ GoalPlannerModule::GoalPlannerModule(
...
     node.create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+  auto lane_parking_planner = std::make_unique<LaneParkingPlanner>(
+    node, lane_parking_mutex_, lane_parking_request_, lane_parking_response_,
+    is_lane_parking_cb_running_, getLogger(), *parameters_);
   lane_parking_timer_ = rclcpp::create_timer(
-    &node, clock_, lane_parking_period_ns, std::bind(&GoalPlannerModule::onTimer, this),
+    &node, clock_, lane_parking_period_ns,
+    [lane_parking_planner_bind = std::move(lane_parking_planner)]() {
+      lane_parking_planner_bind->onTimer();
+    },
```
[the remark this answers is not in the document]
Comment (soblin) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp]: renamed to `lane_parking_executor` in  https://github.com/autowarefoundation/autoware.universe/commit/d42fc0d5ec451e2789e76d8ae5c2cbf52e52becb

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 6]
```diff
@@ -118,19 +91,30 @@ GoalPlannerModule::GoalPlannerModule(
...
       node.create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+    auto freespace_parking_planner = std::make_unique<FreespaceParkingPlanner>(
+      freespace_parking_mutex_, freespace_parking_request_, freespace_parking_response_,
+      is_freespace_parking_cb_running_, getLogger(), clock_, freespace_planner);
     freespace_parking_timer_ = rclcpp::create_timer(
       &node, clock_, freespace_parking_period_ns,
-      std::bind(&GoalPlannerModule::onFreespaceParkingTimer, this),
+      [freespace_parking_planner_bind = std::move(freespace_parking_planner)]() {
+        freespace_parking_planner_bind->onTimer();
+      },
```
[the remark this answers is not in the document]
Comment (soblin) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp]: renamed to `freespace_parking_executor` in  https://github.com/autowarefoundation/autoware.universe/commit/d42fc0d5ec451e2789e76d8ae5c2cbf52e52becb

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 7]
```diff
@@ -613,15 +607,70 @@ void GoalPlannerModule::updateData()
...
+      isStopped(
+        odometry_buffer_stuck_, planner_data_->self_odometry, stuck_time,
+        parameters_->th_stopped_velocity));
     // GoalPlannerModule::occupancy_grid_map_ and gp_planner_data.occupancy_grid_map share the
     // ownership, and gp_planner_data.occupancy_grid_map maybe also shared by the local
     // planner_data on onFreespaceParkingTimer thread local memory space. So following operation
     // is thread-safe because gp_planner_data.occupancy_grid_map is only re-pointed here and its
     // prior resource is still owned by the onFreespaceParkingTimer thread locally.
-    occupancy_grid_map_ = gp_planner_data.occupancy_grid_map;
+    occupancy_grid_map_ = freespace_parking_request.get_occupancy_grid_map();
```
[the remark this answers is not in the document]
Comment (soblin) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp]: remove ogm-based-collision-detector from FreespaceExecutor in the future

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp (hunk 4 above)
Comment (kosuke55) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:662]: by doing
- remove occupancy_grid from isStuck()
- do goal_searcher_->update() in updateData() and passed to lane_parking/freespace_parking thread.

we do not need to share occupnacy grid map.

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 8]
```diff
@@ -1413,59 +1438,64 @@ BehaviorModuleOutput GoalPlannerModule::planPullOverAsCandidate(
...
     auto goal_candidates = goal_candidates_;
     goal_searcher_->update(
       goal_candidates, occupancy_grid_map_, planner_data_, context_data.static_target_objects);
 
     // Select a path that is as safe as possible and has a high priority.
-    const auto & pull_over_path_candidates = context_data.pull_over_path_candidates;
-    auto path_and_goal_opt =
+    const auto & pull_over_path_candidates =
+      context_data.lane_parking_response.pull_over_path_candidates;
+    auto lane_pull_over_path_opt =
```
[the remark this answers is not in the document]
Comment (soblin) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp]: fixed in 181415e65cd71233df5065fec2defb4d7410118b

**Review by kosuke55 — APPROVED** [review]

**Pull request merged**